File: php/Imports/createManualImportOn2Steps.php
<?php

/**
 * Ce script permet de créer un Import Manuel en 2 étapes seulements.
 *
 * @package cookbook
 */

require __DIR__ . '/../utils.php';

/**
 * Variable contenant les configurations pour se connecter à l'API.
 *
 * @var array
 */
$configs = parse_ini_file(__DIR__. '/../config.ini');
/**
 * ID du binding contenant les configurations du fichier à importer.
 *
 * @var int
 */
$idBinding = 1234;

// Fonctions

/**
 * Création du Json contenant les informations de l'import.
 *
 * @param integer $idBinding
 *
 * @return string
 */
function createImportJson($idBinding) {

  // TODO
  $importName = "Manual Import PHP"; //Nom de l'import

  $segmentId = 12345; // Id du segment
  $contactsId = ["XXXXXXX"]; // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
  $groupsContactsId = []; // Id des groupes : Administration -> Groupes -> Idenfitifant

  $importData = array(
    "name"                  => $importName,
    "features"              => array(),
    "binding"               => $idBinding,
  );
  $importData["features"] = [
  array( // Ajout des cibles a un Segment
    "type"                  => "segmentation",
    "segmentId"             => $segmentId,
    "emptyExisitingSegment" => false
    ),
  array( // Regles a appliquer en cas de doublon
    "type"                  => "duplicate",
    "rules"                 => ["ignore" => true]
   ),
  array( // Parametres du rapport
    "type"                  => "report",
    "sendFinalReport"       => true,
    "sendErrorReport"       => true,
    "contactGuids"          => $contactsId, // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
    "groupIds"              => $groupsContactsId, // Id des groupes : Administration -> Groupes -> Idenfitifant
    ),
  array( // Parametres pour mettre a jour une cible dans la base de donnees
    "type"                  => "database",
    "updateExisting"        => true,
    "crushData"             => false
    )
  ];

  $createImportJson = json_encode($importData);

  echo "json :\n" . $createImportJson . "\n";

  return ($createImportJson);
}

/**
 * Création du Json contenant le contenue de l'import.
 *
 *
 * @return array
 */
function getExecutionImportJson() {
  // Remplissez les informations obligatoires
  $fromFile = "./";
  $nameFile = "import.csv";
  $filePath = $fromFile . $nameFile;
  $valid = false;
  $validExt = [
      'txt',
      'csv',
      'zip',
      'tar.gz',
      'tgz',
      'gz'
  ];
  $validExtLength = count($validExt);

  $fileParts = pathinfo($filePath);

  if (!$fileParts || !isset($fileParts['extension']))
  {
      throw new Exception('There was a problem with your file');
  }

  for ($i = 0; $i < $validExtLength; $i++)
  {
      if ($validExt[$i] == $fileParts['extension'])
      {
          $valid = true;
      }
  }

  if (!$valid)
  {
      throw new Exception('Wrong type of file');
  }

  $finfo = finfo_open(FILEINFO_MIME_TYPE);

  $contentType = finfo_file($finfo, $filePath);

  finfo_close($finfo);

  $cfile = curl_file_create($filePath, $contentType, $fileParts['basename']);

  $data = ['data' => $cfile];

  return $data;
}

// On cree l'import de base
echo "CREATION IMPORT\n";
$url = $configs['url'] . "imports";
$con = connect($url, $configs['xKey'], createImportJson($idBinding), 'POST');
if ($con['result'] == false) {
  // Erreur lors de la requete
  // Affichage de l'erreur
  echo "Error : " . $con['info']['http_code'] . "\n" . $con['result'];
  return;
}

$importId = json_decode($con['result'], true)['id'];

// On execute l'import
echo "EXECUTION IMPORT\n";
$url = $configs['url'] . "imports/" . $importId . "/executions";
$con = sendFile($url, $configs['xKey'], getExecutionImportJson(), 'POST');
if ($con['result'] == false) {
  // Erreur lors de la requete
  // Affichage de l'erreur
  echo "Error : " . $con['info']['http_code'] . "\n" . $con['result'];
  return;
}

echo "Result : " . $con['result'] . ":\n";


?>

File: Csharp/Imports/createManualImportOn2Steps.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace createManualImportOn2Steps
{
    class createManualImportOn2Steps
    {
        static JObject getCreateImportJson(int idBinding)
        {
            // Remplissez les informations obligatoires
            string importName = "Nom de votre Import Manuel"; //Nom de l'import
            int segmentId = 1234; // Id du segment
            String phoneFieldsId = "1234"; // Id du field
            String[] contactsId = { "1234ABCD" }; // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
            int[] groupsContactsId = { 1234 }; // Id des groupes : Administration -> Groupes -> Idenfitifant

            JArray phoneFieldsIdArray = new JArray(); // Pour changer de pays de normalisation utiliser l'ISO ALPHA-3 Code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)
            phoneFieldsIdArray.Add("FRA");

            JObject fields = new JObject();
            fields.Add(phoneFieldsId, phoneFieldsIdArray);

            JObject normalization = new JObject(); // Normalisation
            normalization.Add("type", "normalization");
            normalization.Add("fields", fields);

            JObject segmentation = new JObject(); // Ajout des cibles a un Segment
            segmentation.Add("type", "segmentation");
            segmentation.Add("segmentId", segmentId);
            segmentation.Add("emptyExisitingSegment", true);

            JObject destination = new JObject();
            destination.Add("sms", true);
            destination.Add("email", true);

            JObject redList = new JObject(); // Ajout des cibles a la liste rouge
            redList.Add("type", "redList");
            redList.Add("destination", destination);

            JObject rules = new JObject();
            rules.Add("ignore", true); // ou rules.Add("first", true); ou rules.Add("last", true);

            JObject duplicate = new JObject(); // Regles a appliquer en cas de doublon
            duplicate.Add("type", "duplicate");
            duplicate.Add("rules", rules);

            JArray contactGuids = new JArray();
            contactGuids.Add(contactsId);

            JArray groupIds = new JArray();
            groupIds.Add(groupsContactsId);

            JObject report = new JObject(); // Parametres du rapport
            report.Add("type", "report");
            report.Add("sendFinalReport", false);
            report.Add("sendErrorReport", true);
            report.Add("contactGuids", contactGuids); // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
            report.Add("groupIds", groupIds); // Id des groupes : Administration -> Groupes -> Idenfitifant

            JObject database = new JObject(); // Parametres pour mettre a jour une cible dans la base de donnees
            database.Add("type", "database");
            database.Add("updateExisting", true);
            database.Add("crushData", false);

            JArray features = new JArray();
            features.Add(normalization);
            features.Add(segmentation);
            //features.Add(redList);
            features.Add(duplicate);
            features.Add(report);
            features.Add(database);

            JObject createImportJson = new JObject();
            createImportJson.Add("name", importName);
            createImportJson.Add("features", features);
            createImportJson.Add("binding", idBinding);

            //On affiche le Json
            Console.Write("Json :\n" + createImportJson + "\n\n");

            return (createImportJson);
        }

        static string[] getFileSourceInfo()
        {
            // Remplissez les informations obligatoires
            string path = "C:\\votre\\chemin\\vers\\le\\fichier\\";
            string name = "nomDeVotreFichier.extension";
            string file = null;

            System.IO.StreamReader myFile =
                new System.IO.StreamReader(path + name);
            file = myFile.ReadToEnd();

            string[] result = { name, path, file };
            return (result);
        }


        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        /////FIN DE CREATION DES JSONS////////////////////////////////////////////////////////////////////////////////////////////////////
        /////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

        static void Main(string[] args)
        {
            //Ne pas remplir
            Object[] result = null;
            HttpWebRequest con = null;
            HttpWebResponse httpResponse = null;

            //Url de base
            String urlBase = "https://backoffice.mailperformance.com/";
            //X-Key
            String xKey = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
            int idBinding = 1234; //Bindings cree par les APIs : importFormats.

            //Lancement de la connexion pour remplir la requete
            String url = urlBase + "imports/";
            Console.Write("CREATE IMPORT\n");
            result = allConnection(con, url, xKey, getCreateImportJson(idBinding), "POST");
            if (verification(result) == false)
                System.Environment.Exit(0);
            int importId = ((dynamic)(JsonConvert.DeserializeObject((String)result[3]))).id;

            //Nouvelle url
            url = urlBase + "imports/" + importId + "/executions";
            //On execute l'import
            Console.Write("EXECUTION IMPORT\n");
            result = sendFile(con, url, xKey, getFileSourceInfo(), "POST");
            if ((verification(result)) == false)
                System.Environment.Exit(0);

            Console.ReadLine();
            httpResponse = (HttpWebResponse)result[2];
            httpResponse.Close();
        }

        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        /////FIN DU PROGRAMME/////////////////////////////////////////////////////////////////////////////////////////////////////////////
        /////DEBUT DES FONCTIONS//////////////////////////////////////////////////////////////////////////////////////////////////////////
        //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

        static Boolean verification(Object[] result)
        {
            int response = 0;
            HttpWebRequest con = null;
            HttpWebResponse httpResponse = null;
            String responseString = null;

            response = (int)result[0];
            con = (HttpWebRequest)result[1];
            httpResponse = (HttpWebResponse)result[2];
            responseString = (String)result[3];

            if (response >= 200 && response <= 299)
            {
                //On affiche la reponse
                Console.Write("OK : \n\n" + responseString + "\n\n");
                return (true);
            }
            Console.Write("Error : " + response + " " + responseString + "\n");
            httpResponse = (HttpWebResponse)result[2];
            httpResponse.Close();
            Console.ReadLine();
            return (false);
        }

        //Fonction de connexion
        static HttpWebRequest Connect(String url, String xKey, String method)
        {
            //Lancement de la connexion pour remplir la requete
            HttpWebRequest con = (HttpWebRequest)WebRequest.Create(url);
            con.Method = method;

            //Mise en place du xKey et des options
            con.Headers.Add("X-Key", xKey);
            con.ContentType = "application/json";
            con.Accept = "application/vnd.mperf.v8.import.v1+json";

            return (con);
        }

        //Fonction de connexion et envoie des informations
        static Object[] allConnection(HttpWebRequest con, String url, String xKey, JObject jsonMessage, String method)
        {
            con = Connect(url, xKey, method);
            con.ContentLength = 0;
            if (jsonMessage != null)
            {
                con.ContentLength = jsonMessage.ToString().Length;
                var streamWriter = new StreamWriter(con.GetRequestStream());
                streamWriter.Write(jsonMessage);
                streamWriter.Flush();
                streamWriter.Close();
            }

            //Test de l'envoi
            HttpWebResponse httpResponse = null;
            int response = 0;
            String responseString = null;
            try
            {
                httpResponse = (HttpWebResponse)con.GetResponse();
                response = 200;

                //Lecture des donnees
                StreamReader reader = new StreamReader(con.GetResponse().GetResponseStream());
                responseString = reader.ReadToEnd();
                httpResponse.Close();
                reader.Close();
            }
            //Reception du signal
            catch (WebException ex)
            {
                if (ex.Status == WebExceptionStatus.ProtocolError)
                {
                    httpResponse = (HttpWebResponse)ex.Response;
                    response = (int)httpResponse.StatusCode;
                }
            }
            Object[] result = { response, con, httpResponse, responseString };
            return (result);
        }

        static Object[] sendFile(HttpWebRequest con, String url, String xKey, String[] fileInfo, String method)
        {
            string nameFile = fileInfo[0];
            string pathAndNameFile = fileInfo[1] + fileInfo[0];
            string file = fileInfo[2];

            Console.Write("\nnamefile = " + nameFile + "\npathAndNameFile = " + pathAndNameFile + "\nFILE = " + file + "\n");

            //Lancement de la connexion pour remplir la requete
            con = (HttpWebRequest)WebRequest.Create(url);
            con.Method = method;

            //Mise en place du xKey et des options
            con.Headers.Add("X-Key", xKey);
            con.Headers.Add("Content-Disposition", "form-data; filename=" + nameFile);
            con.Accept = "application/vnd.mperf.v8.import.v1+json";
            con.ContentType = "application/octet-stream";

            var streamWriter = new StreamWriter(con.GetRequestStream());
            streamWriter.Write(file);
            streamWriter.Flush();
            streamWriter.Close();

            //Test de l'envoi
            HttpWebResponse httpResponse = null;
            int response = 0;
            String responseString = null;
            try
            {
                httpResponse = (HttpWebResponse)con.GetResponse();
                response = 200;

                //Lecture des donnees
                StreamReader reader = new StreamReader(con.GetResponse().GetResponseStream());
                responseString = reader.ReadToEnd();
                httpResponse.Close();
                reader.Close();
            }
            //Reception du signal
            catch (WebException ex)
            {
                if (ex.Status == WebExceptionStatus.ProtocolError)
                {
                    httpResponse = (HttpWebResponse)ex.Response;
                    response = (int)httpResponse.StatusCode;
                }
            }
            Object[] result = { response, con, httpResponse, responseString };
            return (result);
        }
    }
}
